File: examples/C/simple/simple.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <wchar.h>
#include <time.h>
#include <string.h>
#include <stdlib.h>
#include <hidapi.h>
#include <simple.h>
#include <buddy.h>
#include <usbhid_buddy.h>
#include <support.h>
#include <utility.h>

#include <sys/timeb.h>

extern char *fw_info_dac_type_names[FIRMWARE_INFO_DAC_TYPE_LENGTH];

static hid_device* hid_handle;
static bool active = false;

#if defined(_WIN32) || defined(_WIN64)
#include <windows.h>

BOOL my_exit_handler(DWORD fdwCtrlType)
{
	switch (fdwCtrlType)
	{
		case CTRL_C_EVENT:
			printf("Ctrl-C detected\r\n");
			active = false;
			return TRUE;

		default:
			break;
	}

	return FALSE;
}
#else
#include <signal.h>
void my_exit_handler(int s)
{
	printf("my_exit_handler caught\n");
	exit(1);
}
#endif

void display_usb_info(buddy_hid_info_t *hid_info)
{
	printf("\nUSB Manufacturer String: %s\n", hid_info->str_mfr);
	printf("USB Product String: %s\n", hid_info->str_product);
	printf("USB Serial Number String: %s\n", hid_info->str_serial);
	printf("USB Indexed String 1: %s\n\n", hid_info->str_index_1);
}

void display_fw_info(firmware_info_t *fw_info)
{
	char res[128];
	struct tm *timeinfo;
	time_t flashtime;

	printf("Firmware Serial Number: %d (0x%x)\n", 
		fw_info->serial, fw_info->serial);
	printf("Firmware Revision: %d.%d.%d\n",
		fw_info->fw_rev_major, fw_info->fw_rev_minor, fw_info->fw_rev_tiny);
	printf("Firmware Bootload Revision: %d.%d.%d\n",
		fw_info->bootl_rev_major, fw_info->bootl_rev_minor, fw_info->bootl_rev_tiny);
	
	flashtime = fw_info->flash_datetime;
	timeinfo = localtime( &flashtime );

	if (strftime(res, sizeof(res), "%Y-%m-%d %H:%M:%S", timeinfo) > 0) {
		printf("Firmware Flash datetime: %s\n", res);
	} else {
		printf("Firmware Flash datetime: none\n");
	}

	if (fw_info->type_dac < FIRMWARE_INFO_DAC_TYPE_LENGTH) {
		printf("Firmware DAC type: %d - %s\n",
			fw_info->type_dac, fw_info_dac_type_names[fw_info->type_dac]);
	} else {
		printf("Firmware DAC type: %d\n",
			fw_info->type_dac);
	}
}

int8_t test_seq_pwm_freq(hid_device* handle, firmware_info_t *fw_info,
	float sample_rate, bool streaming)
{
	static int test_seq_pwm_count = 0;
	general_packet_t packet;
	ctrl_general_t general_settings = { 0 };
	ctrl_timing_t timing_settings = { 0 };
	ctrl_runtime_t runtime_settings = { 0 };
	int k;
	int i;
	int f;

	active = true;
	general_settings.function = GENERAL_CTRL_PWM_ENABLE;
	general_settings.mode = (streaming ? MODE_CTRL_STREAM : MODE_CTRL_IMMEDIATE);
	general_settings.channel_mask = BUDDY_CHAN_0_MASK;
	general_settings.resolution = RESOLUTION_CTRL_SUPER;
	//general_settings.resolution = RESOLUTION_CTRL_HIGH;
	//general_settings.resolution = RESOLUTION_CTRL_LOW;

	timing_settings.period = (uint32_t)FREQUENCY_TO_NSEC(sample_rate);

	runtime_settings.pwm_mode = RUNTIME_PWM_MODE_FREQUENCY;
	runtime_settings.pwm_timebase = RUNTIME_PWM_TIMEBASE_SYSCLK_DIV_12;

	if (buddy_configure(handle, &general_settings, &runtime_settings, &timing_settings) != BUDDY_ERROR_CODE_OK) {
		printf("test_seq_pwm_freq: could not buddy_init\n");
		return -1;
	}
	short_sleep(100);

	//for (k = 20000; k <= 20000; k++) {
	for (k = 100000; k <= 100000; k++) {
		if (!active) {
			return 0;
		}

		for (i = BUDDY_CHAN_0; i <= BUDDY_CHAN_7; i++) {
			packet.channels[i] = k;
		}

		printf("test_seq_pwm_freq, sending %d packet with value %d (%x)\r\n", test_seq_pwm_count++, k, k);

		if (buddy_send_pwm(handle, &packet, streaming) != BUDDY_ERROR_CODE_OK) {
			printf("test_seq_pwm_freq: buddy_send_pwm call failed\n");
			return -1;
		}

		// if streaming is not enabled then wait explicitly but otherwise
		// call to send the next packet and let the encoder work
		if (!streaming) {
			short_sleep((int)((1.0 / sample_rate) / 1e-3));
		}
	}

	short_sleep(10);
	// flush the stream frame buffer if packets are in it but did not fill whole frame
	// to send packet, then wait fixed period to allow MCU to process frame
	buddy_flush(handle);

	return 0;
}

int8_t test_seq_pwm_duty(hid_device* handle, firmware_info_t *fw_info,
	float sample_rate, bool streaming)
{
	static int test_seq_pwm_count = 0;
	general_packet_t packet;
	ctrl_general_t general_settings = { 0 };
	ctrl_timing_t timing_settings = { 0 };
	ctrl_runtime_t runtime_settings = { 0 };
	int k;
	int i;
	int f;

	active = true;
	general_settings.function = GENERAL_CTRL_PWM_ENABLE;
	general_settings.mode = (streaming ? MODE_CTRL_STREAM : MODE_CTRL_IMMEDIATE);
	general_settings.channel_mask = BUDDY_CHAN_0_MASK;
	general_settings.resolution = RESOLUTION_CTRL_HIGH;
	//general_settings.resolution = RESOLUTION_CTRL_LOW;

	timing_settings.period = (uint32_t)FREQUENCY_TO_NSEC(sample_rate);

	runtime_settings.pwm_mode = RUNTIME_PWM_MODE_DUTY_CYCLE;
	runtime_settings.pwm_timebase = RUNTIME_PWM_TIMEBASE_SYSCLK;

	if (buddy_configure(handle, &general_settings, &runtime_settings, &timing_settings) != BUDDY_ERROR_CODE_OK) {
		printf("test_seq_pwm_duty: could not buddy_init\n");
		return -1;
	}
	short_sleep(100);

	for (k = 32767; k <= 32767; k++) {
	//for (k = 16383; k <= 16383; k++) {
	//for (k = 128; k <= 128; k++) {
		if (!active) {
			return 0;
		}

		for (i = BUDDY_CHAN_0; i <= BUDDY_CHAN_7; i++) {
			packet.channels[i] = k;
		}

		printf("test_seq_pwm_duty, sending %d packet with value %d (%x)\r\n", test_seq_pwm_count++, k, k);

		if (buddy_send_pwm(handle, &packet, streaming) != BUDDY_ERROR_CODE_OK) {
			printf("test_seq_pwm_duty: buddy_send_pwm call failed\n");
			return -1;
		}

		// if streaming is not enabled then wait explicitly but otherwise
		// call to send the next packet and let the encoder work
		if (!streaming) {
			short_sleep((int)((1.0 / sample_rate) / 1e-3));
		}
	}

	short_sleep(10);
	// flush the stream frame buffer if packets are in it but did not fill whole frame
	// to send packet, then wait fixed period to allow MCU to process frame
	buddy_flush(handle);

	return 0;
}

int8_t test_seq_dac(hid_device* handle, firmware_info_t *fw_info, 
				float sample_rate, bool streaming)
{
	static int test_seq_dac_count = 0;
	general_packet_t packet;
	ctrl_general_t general_settings = { 0 };
	ctrl_timing_t timing_settings = { 0 };
	ctrl_runtime_t runtime_settings = { 0 };
	int k;
	int i;
	int f;

	active = true;
	general_settings.function = GENERAL_CTRL_DAC_ENABLE;
	general_settings.mode = (streaming ? MODE_CTRL_STREAM : MODE_CTRL_IMMEDIATE);
	//general_settings.channel_mask = BUDDY_CHAN_ALL_MASK;
	general_settings.channel_mask = BUDDY_CHAN_0_MASK;
	general_settings.resolution = RESOLUTION_CTRL_HIGH;
	//general_settings.resolution = RESOLUTION_CTRL_LOW;

	timing_settings.period = (uint32_t) FREQUENCY_TO_NSEC(sample_rate);
	
	runtime_settings.dac_power = RUNTIME_DAC_POWER_ON;
	runtime_settings.dac_ref = RUNTIME_DAC_REF_EXT;
	//runtime_settings.dac_ref = RUNTIME_DAC_REF_INT_1V;
	//runtime_settings.dac_ref = RUNTIME_DAC_REF_INT_2V;

	if 	(buddy_configure(handle, &general_settings, &runtime_settings, &timing_settings) != BUDDY_ERROR_CODE_OK) {
		printf("test_seq_dac: could not buddy_init\n");
		return -1;
	}
	short_sleep(100);

	for (k = 0; k <= 4095; k++) {
		if (!active) {
			return 0;
		}
		
		for (i = BUDDY_CHAN_0; i <= BUDDY_CHAN_7; i++) {
			packet.channels[i] = k;
		}

		//printf("test_seq_dac, sending %d packet with value %d\r\n", test_seq_dac_count++, (k % 255));
		printf("test_seq_dac, sending %d packet with value %d (%x)\r\n", test_seq_dac_count++, k, k);

		if (buddy_send_dac(handle, &packet, streaming) != BUDDY_ERROR_CODE_OK) {
			printf("test_seq_dac: buddy_send_dac call failed\n");
			return -1;
		}

		// if streaming is not enabled then wait explicitly but otherwise
		// call to send the next packet and let the encoder work
		if (!streaming) {
			short_sleep( (int) ((1.0 / sample_rate) / 1e-3));
		} 
	}

	short_sleep(10);
	// flush the stream frame buffer if packets are in it but did not fill whole frame
	// to send packet, then wait fixed period to allow MCU to process frame
	buddy_flush(handle);

	return 0;
}

int8_t test_seq_adc(hid_device* handle, firmware_info_t *fw_info,
				float sample_rate, bool streaming) {
	static int test_seq_dac_count = 0;
	general_packet_t packet;
	ctrl_general_t general_settings = { 0 };
	ctrl_timing_t timing_settings = { 0 };
	ctrl_runtime_t runtime_settings = { 0 };
	int i;
	int err_code;
	int recv_packets;

	active = true;
	general_settings.function = GENERAL_CTRL_ADC_ENABLE;
	general_settings.mode = (streaming ? MODE_CTRL_STREAM : MODE_CTRL_IMMEDIATE);
	//general_settings.channel_mask = BUDDY_CHAN_ALL_MASK;
	//general_settings.channel_mask = BUDDY_CHAN_6_MASK | BUDDY_CHAN_0_MASK | BUDDY_CHAN_7_MASK;
	general_settings.channel_mask = BUDDY_CHAN_2_MASK;
	general_settings.resolution = RESOLUTION_CTRL_HIGH;
	//general_settings.resolution = RESOLUTION_CTRL_LOW;

	timing_settings.period = (uint32_t) FREQUENCY_TO_NSEC(sample_rate);
	timing_settings.averaging = 1;

	runtime_settings.adc_ref = RUNTIME_ADC_REF_VDD;

	printf("timing_settings.period = %d (0x%x)\n", timing_settings.period, timing_settings.period);

	if 	(buddy_configure(handle, &general_settings, &runtime_settings, &timing_settings) != BUDDY_ERROR_CODE_OK) {
		printf("test_seq_dac: could not buddy_init\n");
		return -1;
	}
	short_sleep(100);

	recv_packets = 0;
	do {
		if (!active) {
			return 0;
		}

		err_code = buddy_read_adc(handle, &packet, streaming);
		if (err_code == BUDDY_ERROR_CODE_OK) {
			printf("test_seq_adc: received packet %d\n", recv_packets);

			for (i = BUDDY_CHAN_0; i <= BUDDY_CHAN_7; i++) {
				if (general_settings.channel_mask & (1 << i)) {
					printf("packet.channels[%d] = %d\r\n", i, packet.channels[i]);
				}
			}

			recv_packets++;
		} else if (err_code == BUDDY_ERROR_CODE_GENERAL) {
			printf("test_seq_adc: buddy_send_adc call failed\n");
			return BUDDY_ERROR_CODE_GENERAL;
		}
	} while (recv_packets < 500);

	return 0;
}

int main(int argc, char *argv[])
{
	buddy_hid_info_t hid_info;
	firmware_info_t fw_info;
	hid_device *hid_handle;

	int optind;
	int8_t err_code = 0;
	struct timeb time_start, time_end;
	int time_diff;
	bool stream_mode = false;
	bool hold_mode = false;
	enum { DAQ_MODE, ADC_MODE, PWM_DUTY_MODE, PWM_FREQ_MODE } mode = DAQ_MODE;

	#ifdef WIN32
	UNREFERENCED_PARAMETER(argc);
	UNREFERENCED_PARAMETER(argv);
	#endif
	
	#if defined(_WIN32) || defined(_WIN64)
	if (!SetConsoleCtrlHandler((PHANDLER_ROUTINE)my_exit_handler, TRUE))
	{
		printf("\nERROR: Could not set control handler\r\n");
		return 1;
	}
	#else
	struct sigaction sigIntHandler;
	sigIntHandler.sa_handler = my_exit_handler;
	sigemptyset(&sigIntHandler.sa_mask);
	sigIntHandler.sa_flags = 0;

	sigaction(SIGINT, &sigIntHandler, NULL);
	#endif

    for (optind = 1; optind < argc; optind++) {
		switch (argv[optind][1]) {
			case 'd': mode = DAQ_MODE; break;
			case 'a': mode = ADC_MODE; break;
			case 'p': mode = PWM_DUTY_MODE; break;
			case 't': mode = PWM_FREQ_MODE; break;
			case 's': stream_mode = true; break;
			case 'l': hold_mode = true; break;
			case 'h': 
			default:
				fprintf(stdout, "Usage: %s [-daptls]\n", argv[0]);
				fprintf(stdout, "\nTest the DAC, ADC, and PWM function of the Buddy DAQ device\n\n");
				fprintf(stdout, "optional arguments:\n");
				fprintf(stdout, " -h,        show this help message and exit\n");
				fprintf(stdout, " -d,        enable DAC(data to analog) mode\n");
				fprintf(stdout, " -a,        enable ADC(analog to digital) mode\n");
				fprintf(stdout, " -p,        enable PWM(pulse width modulation) duty cycle mode\n");
				fprintf(stdout, " -t,        enable PWM(pulse width modulation) frequency mode\n");
				fprintf(stdout, " -s,        enable streaming for higher throughput\n");
				fprintf(stdout, " -l,        hold the last mode and value on exit\n");
				exit(EXIT_FAILURE);
        } 
    }

	// open Buddy HID device and retrieve the USB device and firmware info
	hid_handle = buddy_init(&hid_info, &fw_info);

	if (!hid_handle) {
		printf("main: could not open Buddy HID device\n");
		return -1;
	}

	display_usb_info(&hid_info);
	display_fw_info(&fw_info);

    // check if revision is acceptable
    if ((fw_info.fw_rev_major != 0) || (fw_info.fw_rev_minor != 4)) {
        printf("main: firmware major/minor version is not valid, exiting.\n");
        buddy_cleanup(hid_handle, &hid_info, false);
        return -1;
    }

	ftime(&time_start);
	if (mode == DAQ_MODE) {
		printf("main: testing with mode = DAQ_MODE\n");
		err_code = test_seq_dac(hid_handle, &fw_info, BUDDY_TEST_DAC_FREQ, stream_mode);
	} else if (mode == ADC_MODE) {
		printf("main: testing with mode = ADC_MODE\n");
		err_code = test_seq_adc(hid_handle, &fw_info, BUDDY_TEST_ADC_FREQ, stream_mode);
	} else if (mode == PWM_DUTY_MODE) {
		printf("main: testing with mode = PWM_DUTY_MODE\n");
		err_code = test_seq_pwm_duty(hid_handle, &fw_info, BUDDY_TEST_PWM_FREQ, stream_mode);
	} else if (mode == PWM_FREQ_MODE) {
		printf("main: testing with mode = PWM_FREQ_MODE\n");
		err_code = test_seq_pwm_freq(hid_handle, &fw_info, BUDDY_TEST_PWM_FREQ, stream_mode);
	}

	ftime(&time_end);
	time_diff = (int)(1000.0 * (time_end.time - time_start.time) + (time_end.millitm - time_start.millitm));
	printf("Test took (%f) seconds  to run\n", (float) (time_diff / 1000.0));
	
	short_sleep(1000);

	if (err_code == 0) {
		printf("hold_mode = %d\n", hold_mode);

		if (hold_mode) {
			buddy_cleanup(hid_handle, &hid_info, false);
		} else {
			buddy_cleanup(hid_handle, &hid_info, true);
		}
	} else {
		printf("main: could not complete requested buddy action\n");
	}
	
	return 0;
}

File: lib/libbuddy/bindings/python/simple.py
#!/usr/bin/env python

import buddy as bt
import sys
import datetime
import argparse
import time
import signal

BUDDY_TEST_ADC_FREQ = 1000      # 1 kHz
BUDDY_TEST_DAC_FREQ = 1000      # 5 Hz

hid_handle = None
hid_info = None

def test_seq_dac(handle, fw_info, sample_rate, streaming, oneshot):
    general_settings = bt.ctrl_general_t()
    timing_settings = bt.ctrl_timing_t()
    runtime_settings = bt.ctrl_runtime_t()

    general_settings.function = bt.GENERAL_CTRL_DAC_ENABLE
    general_settings.mode = \
        bt.MODE_CTRL_STREAM if streaming else bt.MODE_CTRL_IMMEDIATE
    general_settings.operation = \
        bt.OPER_CTRL_ONESHOT if oneshot else bt.OPER_CTRL_CONTINUOUS
    general_settings.queue = bt.QUEUE_CTRL_WAIT
    general_settings.channel_mask = bt.BUDDY_CHAN_ALL_MASK
    #general_settings.channel_mask = bt.BUDDY_CHAN_0_MASK
    #general_settings.channel_mask = bt.BUDDY_CHAN_0_MASK | bt.BUDDY_CHAN_1_MASK
    general_settings.resolution = bt.CODEC_BIT_WIDTH_12

    timing_settings.period = bt.FREQUENCY_TO_NSEC(sample_rate)

    runtime_settings.dac_power = bt.RUNTIME_DAC_POWER_ON
    runtime_settings.dac_ref = bt.RUNTIME_DAC_REF_EXT

    if (bt.buddy_configure(handle,
                           general_settings,
                           runtime_settings,
                           timing_settings) != bt.BUDDY_ERROR_OK):
        print 'test_seq_dac: could not configure Buddy device'
        return -1

    time.sleep(0.1)

    packet = bt.general_packet_t()
    test_seq_dac_count = 0

    for k in range(0, ((1 << general_settings.resolution) - 1)):
        for i in range(bt.BUDDY_CHAN_0, bt.BUDDY_CHAN_7):
            bt.uint16_t_ptr_setitem(packet.channels, i, k)

        print 'test_seq_dac, sending %d packet with value %d' % \
                (test_seq_dac_count, k)
        test_seq_dac_count += 1

        if (bt.buddy_send_dac(handle,
                              packet,
                              streaming) != bt.BUDDY_ERROR_OK):
            print 'test_seq_dac: could not send DAC packet'
            return -1

        if not streaming:
            time.sleep(1.0 / sample_rate)

    if not bt.codec_buffer_empty():
        print 'test_seq_dac: flushing the buffer'
        bt.buddy_flush(handle)
    time.sleep(0.1)

    if oneshot:
        if bt.buddy_trigger(handle) != bt.BUDDY_ERROR_OK:
            return -1
        time.sleep(0.1)

    return 0

def test_seq_adc(handle, fw_info, sample_rate, streaming, oneshot):
    general_settings = bt.ctrl_general_t()
    timing_settings = bt.ctrl_timing_t()
    runtime_settings = bt.ctrl_runtime_t()
    packet = bt.general_packet_t()

    general_settings.function = bt.GENERAL_CTRL_ADC_ENABLE
    general_settings.mode = \
        bt.MODE_CTRL_STREAM if streaming else bt.MODE_CTRL_IMMEDIATE
    general_settings.operation = \
        bt.OPER_CTRL_ONESHOT if oneshot else bt.OPER_CTRL_CONTINUOUS
    general_settings.queue = bt.QUEUE_CTRL_SATURATE
    general_settings.channel_mask = bt.BUDDY_CHAN_ALL_MASK
    #general_settings.channel_mask = bt.BUDDY_CHAN_6_MASK
    general_settings.resolution = bt.CODEC_BIT_WIDTH_12

    timing_settings.period = bt.FREQUENCY_TO_NSEC(sample_rate)
    timing_settings.averaging = 1

    runtime_settings.adc_ref = bt.RUNTIME_ADC_REF_VDD

    print 'timing_settings.period = %d (0x%x)' % (timing_settings.period, timing_settings.period)

    if (bt.buddy_configure(handle,
                           general_settings,
                           runtime_settings,
                           timing_settings) != bt.BUDDY_ERROR_OK):
        print 'test_seq_adc: could not configure Buddy device'
        return -1

    time.sleep(0.1)

    if oneshot:
        if bt.buddy_trigger(handle) != bt.BUDDY_ERROR_OK:
            return -1
        time.sleep(0.1)

    recv_packets = 0;
    for i in range(0, 1200):
        err_code = bt.buddy_read_adc(handle, packet, streaming)

        if err_code == bt.BUDDY_ERROR_OK:
            print 'test_seq_adc: received packet %d' % recv_packets
            for j in range(bt.BUDDY_CHAN_0, bt.BUDDY_CHAN_7 + 1):
                if (general_settings.channel_mask & (1 << j)):
                    print 'packet.channels[%d] = %d' % \
                        (j, bt.uint16_t_ptr_getitem(packet.channels, j))

            recv_packets += 1
        elif err_code == bt.BUDDY_ERROR_GENERAL:
            print 'test_seq_adc: could not read ADC packet'
            print 'err_code = %d' % err_code
            return -1
    
    return 0;

def display_usb_info(hid_info):
    print 'USB Manufacturer String: %s' % hid_info.str_mfr
    print 'USB Product String: %s' % hid_info.str_product
    print 'USB Serial Number String: %s' % hid_info.str_serial
    print 'USB Indexed String 1: %s' % hid_info.str_index_1
    print ''

def display_fw_info(fw_info):
    print 'Firmware Serial Number: %d (0x%x)' % \
        (fw_info.serial, fw_info.serial)
    print 'Firmware Revision: %d.%d.%d' % \
        (fw_info.fw_rev_major, fw_info.fw_rev_minor, fw_info.fw_rev_tiny)
    print 'Firmware Bootload Revision: %d.%d.%d' % \
        (fw_info.bootl_rev_major, fw_info.bootl_rev_minor, fw_info.bootl_rev_tiny)

    print 'Firmware Flash datetime: %s' % \
        datetime.datetime.fromtimestamp(fw_info.flash_datetime).strftime('%Y-%m-%d %H:%M:%S')

    if (fw_info.type_dac < bt.FIRMWARE_INFO_DAC_TYPE_LENGTH):
        print 'Firmware DAC type: %d - %s' % \
            (fw_info.type_dac,
             bt.char_ptr_getitem(bt.cvar.fw_info_dac_type_names, fw_info.type_dac))
    else:
        print 'Firmware DAC type: %d' % fw_info.type_dac

    if (fw_info.type_ext_memory < bt.FIRMWARE_INFO_MEM_TYPE_LENGTH):
        print 'Firmware External Memory type: %d - %s' % \
            (fw_info.type_ext_memory,
             bt.char_ptr_getitem(bt.cvar.fw_info_mem_type_names, fw_info.type_ext_memory))
    else:
        print 'Firmware External Memory type: %d' % fw_info.type_ext_memory

    print ''


def signal_handler(signal, frame):
    print('You pressed Ctrl+C!')
    bt.buddy_cleanup(hid_handle, hid_info)
    sys.exit(0)

if __name__ == '__main__':
    signal.signal(signal.SIGINT, signal_handler)

    parser = argparse.ArgumentParser(description='Test the DAC and ADC function'
                                                 'of the Buddy DAQ device')

    parser.add_argument('-d,--dac', action='store_true', dest='dac_mode',
                        help='enable DAC (data to analog) conversion')
    parser.add_argument('-a,--adc', action='store_true', dest='adc_mode',
                        help='enable ADC (analog to digital) conversion')
    parser.add_argument('-p,--dual', action='store_true', dest='dual_mode',
                        help='enable dual DAC and ADC conversion')
    parser.add_argument('-s,--stream', action='store_true', dest='stream_mode',
                        help='enable streaming for higher throughput')
    parser.add_argument('-o,--oneshot', action='store_true', dest='oneshot_mode',
                        help='enable oneshot mode for triggered conversion')
    args = parser.parse_args()

    if args.dac_mode and args.adc_mode:
        print 'ERROR: must select either DAC or ADC mode, not both'
        sys.exit()

    if (not args.dac_mode) and (not args.adc_mode) and (not args.dual_mode):
        print 'ERROR: must select DAC, ADC, or dual mode'
        sys.exit()

    hid_info = bt.buddy_hid_info_t()
    fw_info = bt.firmware_info_t()

    hid_handle = bt.buddy_init(hid_info, fw_info)

    if hid_handle is None:
        print 'could not open Buddy HID device'
        sys.exit()

    display_usb_info(hid_info)
    display_fw_info(fw_info)

    time_start = time.time()

    if args.dac_mode:
        print 'running a DAC mode test'
        err_code = test_seq_dac(hid_handle,
                                fw_info,
                                BUDDY_TEST_DAC_FREQ,
                                args.stream_mode,
                                args.oneshot_mode)

    if args.adc_mode:
        print 'running a ADC mode test'
        err_code = test_seq_adc(hid_handle,
                                fw_info,
                                BUDDY_TEST_ADC_FREQ,
                                args.stream_mode,
                                args.oneshot_mode)

    time_end = time.time()
    time_diff = time_end - time_start
    print 'Test took (%f) seconds  to run' % time_diff

    bt.buddy_cleanup(hid_handle, hid_info)